File: packages/snow_draw_core/lib/draw/edit/core/standard_finish_mixin.dart
import '../../models/draw_state.dart';
import '../../models/interaction_state.dart';
import '../../models/selection_overlay_state.dart';
import '../../types/edit_context.dart';
import '../../types/edit_transform.dart';
import '../apply/edit_apply.dart';
import '../preview/edit_preview.dart';
import 'edit_computed_result.dart';
import 'edit_operation.dart';
import 'edit_operation_helpers.dart';

/// Eliminates the duplicated finish/preview boilerplate in edit operations.
///
/// Subclasses implement [computeResult] for operation-specific geometry
/// and [updateOverlay] for operation-specific multi-select overlay
/// updates. The mixin handles the shared commit and preview logic.
///
/// Operations that need different geometry for finish vs. preview (e.g.
/// arrow point deletion on commit) can override [computeFinishResult].
/// When not overridden, [finish] delegates to [computeResult].
mixin StandardFinishMixin on EditOperation {
  /// Computes the geometry result for this operation.
  ///
  /// Returns `null` when the transform is identity or validation fails,
  /// signaling that no changes should be applied.
  ///
  /// Used by [buildPreview] and, unless [computeFinishResult] is
  /// overridden, also by [finish].
  EditComputedResult? computeResult({
    required DrawState state,
    required EditContext context,
    required EditTransform transform,
  });

  /// Computes the geometry result specifically for [finish].
  ///
  /// Override this when the committed result differs from the preview
  /// (e.g. point deletion that only happens on commit). The default
  /// delegates to [computeResult].
  EditComputedResult? computeFinishResult({
    required DrawState state,
    required EditContext context,
    required EditTransform transform,
  }) => computeResult(state: state, context: context, transform: transform);

  /// Returns the updated [SelectionOverlayState] after committing.
  ///
  /// Only called for multi-select contexts. Single-select operations
  /// pass through the current overlay unchanged.
  SelectionOverlayState updateOverlay({
    required SelectionOverlayState current,
    required EditComputedResult result,
    required EditContext context,
  });

  @override
  DrawState finish({
    required DrawState state,
    required EditContext context,
    required EditTransform transform,
  }) {
    final result = computeFinishResult(
      state: state,
      context: context,
      transform: transform,
    );
    if (result == null) {
      return state.copyWith(application: state.application.toIdle());
    }

    final newElements = EditApply.replaceElementsById(
      elements: state.domain.document.elements,
      replacementsById: result.updatedElements,
    );

    final overlay = context.isMultiSelect
        ? updateOverlay(
            current: state.application.selectionOverlay,
            result: result,
            context: context,
          )
        : state.application.selectionOverlay;

    final nextDomain = state.domain.copyWith(
      document: state.domain.document.copyWith(elements: newElements),
    );
    final nextApplication = state.application.copyWith(
      interaction: const IdleState(),
      selectionOverlay: overlay,
    );

    return state.copyWith(domain: nextDomain, application: nextApplication);
  }

  @override
  EditPreview buildPreview({
    required DrawState state,
    required EditContext context,
    required EditTransform transform,
  }) {
    final result = computeResult(
      state: state,
      context: context,
      transform: transform,
    );
    if (result == null) {
      return EditPreview.none;
    }

    return buildEditPreview(
      state: state,
      context: context,
      previewElementsById: result.updatedElements,
      multiSelectBounds: result.multiSelectBounds,
      multiSelectRotation: result.multiSelectRotation,
    );
  }
}

File: packages/snow_draw_rust_engine/src/draw/edit/core/standard_finish_mixin.rs
#![allow(dead_code)]

use std::collections::HashMap;

use crate::draw::edit::apply::edit_apply::EditApply;
use crate::draw::edit::preview::edit_preview::{build_selection_preview, EditPreview};
use crate::draw::models::application_state::SelectionOverlayState;
use crate::draw::models::draw_state::{DomainState, DrawState};
use crate::draw::models::element_state::ElementState;
use crate::draw::types::edit_context::EditContext;
use crate::draw::types::edit_transform::EditTransform;

use super::edit_computed_result::EditComputedResult;
use super::edit_operation::EditOperation;

/// Shared finish/preview behavior for standard edit operations.
///
/// Rust does not support Dart-style mixins, so this trait provides the same
/// reusable workflow that operations can delegate to from their
/// [`EditOperation::finish`] and [`EditOperation::build_preview`] implementations.
pub trait StandardFinishMixin: EditOperation {
    /// Computes the geometry result for this operation.
    ///
    /// Returns `None` when the transform is identity or validation fails.
    fn compute_result(
        &self,
        state: &DrawState,
        context: &EditContext,
        transform: &EditTransform,
    ) -> Option<EditComputedResult>;

    /// Computes the geometry result specifically for commit.
    ///
    /// Override when commit output differs from preview output.
    fn compute_finish_result(
        &self,
        state: &DrawState,
        context: &EditContext,
        transform: &EditTransform,
    ) -> Option<EditComputedResult> {
        self.compute_result(state, context, transform)
    }

    /// Returns the updated overlay after committing.
    ///
    /// Called only for multi-select contexts.
    fn update_overlay(
        &self,
        current: SelectionOverlayState,
        result: &EditComputedResult,
        context: &EditContext,
    ) -> SelectionOverlayState;

    /// Applies computed element updates to domain state.
    fn apply_result_to_domain(
        &self,
        state: &DrawState,
        result: &EditComputedResult,
    ) -> DomainState {
        if result.updated_elements.is_empty() {
            return state.domain.clone();
        }

        let next_elements = EditApply::replace_elements_by_id(
            state.domain.document.elements.clone(),
            &result.updated_elements,
        );
        let next_elements = EditApply::reorder_elements_by_id_order(
            next_elements,
            result.ordered_element_ids.as_deref(),
        );
        let next_document = state
            .domain
            .document
            .copy_with(Some(next_elements), None, None);
        state.domain.copy_with(Some(next_document), None)
    }

    /// Shared commit flow equivalent to Dart `StandardFinishMixin.finish`.
    fn finish_standard(
        &self,
        state: &DrawState,
        context: &EditContext,
        transform: &EditTransform,
    ) -> DrawState {
        let application = state.application.to_idle();
        let result = self.compute_finish_result(state, context, transform);

        let Some(result) = result else {
            return state.copy_with(None, Some(application));
        };

        let overlay = if context.is_multi_select() {
            self.update_overlay(application.selection_overlay.clone(), &result, context)
        } else {
            application.selection_overlay.clone()
        };

        let domain = self.apply_result_to_domain(state, &result);
        let application = application.copy_with(None, None, Some(overlay));

        state.copy_with(Some(domain), Some(application))
    }

    /// Shared preview flow equivalent to Dart `StandardFinishMixin.buildPreview`.
    fn build_preview_standard(
        &self,
        state: &DrawState,
        context: &EditContext,
        transform: &EditTransform,
    ) -> EditPreview {
        let result = self.compute_result(state, context, transform);

        let Some(result) = result else {
            return EditPreview::default();
        };

        self.build_preview_from_result(state, context, &result)
    }

    /// Converts a computed result into an `EditPreview`.
    ///
    /// Override this when the concrete preview type is fully translated.
    fn build_preview_from_result(
        &self,
        state: &DrawState,
        context: &EditContext,
        result: &EditComputedResult,
    ) -> EditPreview {
        build_edit_preview(
            state,
            context,
            &result.updated_elements,
            result.multi_select_bounds,
            result.multi_select_rotation,
        )
    }
}

/// Builds an `EditPreview` from computed preview inputs.
pub fn build_edit_preview(
    state: &DrawState,
    context: &EditContext,
    preview_elements_by_id: &HashMap<String, ElementState>,
    multi_select_bounds: Option<crate::draw::types::draw_rect::DrawRect>,
    multi_select_rotation: Option<f64>,
) -> EditPreview {
    let selection_preview = build_selection_preview(
        state,
        context,
        preview_elements_by_id,
        multi_select_bounds,
        multi_select_rotation,
    );

    EditPreview::new(preview_elements_by_id.clone(), selection_preview)
}
